Add a command for printing rasterized text

nickcharlton opened the pull request:
This allows us to print a line with a custom font, sizes and weights
which resolves problems with printing text of unsupported sizes (#132)
and unsupported character sets/code pages (#116)

We achieve this by rendering an image with text, sized to fit the text
itself and returning `byte[]` to be compatible with other commands.

Fixes #132, #116

---

I'm opening this as a draft, as I'm a bit stuck on testing and some feedback on the API itself would be great. I'm using this in production currently and it's working quite well.

On the project which this is extracted from, I'm comparing the rendered text to a fixture using [a library I cobbled together from some example code](https://github.com/nickcharlton/image-comparer). It uses `System.Drawing.Common`, rather than `ImageSharp` so one option would be for me to refactor that then use it here, but I'm open to any thoughts on that.

Another problem with testing is needing a `ttf` font. A good choice here might be to bundle an open source font for this purpose, but I'm open to ideas.

Finally, it's in the command emitters for images, which felt the most appropriate for what it's doing but it could easily go somewhere else.

nickcharlton committed:
Add a command for printing rasterized text

This allows us to print a line with a custom font, sizes and weights
which resolves problems with printing text of unsupported sizes (#132)
and unsupported character sets/code pages (#116)

We achieve this by rendering an image with text, sized to fit the text
itself and returning `byte[]` to be compatible with other commands.

Fixes #132, #116

rakeshpetit commented on ESCPOS_NET/ESCPOS_NET.csproj:
```diff
@@ -24,7 +24,8 @@ Allows creating receipts with all common functionality supported.</Description>
     <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="5.0.0" />
-    <PackageReference Include="SixLabors.ImageSharp" Version="1.0.0" />
+    <PackageReference Include="SixLabors.ImageSharp" Version="1.0.4" />
+    <PackageReference Include="SixLabors.ImageSharp.Drawing" Version="1.0.0-beta13" />
```
Just adding `SixLabors.ImageSharp` won't get us `SixLabors.ImageSharp.Drawing`? Without C# knowledge, I am asking this.

nickcharlton replied:
[`Drawing` is a separate package](https://www.nuget.org/packages/SixLabors.ImageSharp.Drawing/1.0.0-beta14).

rakeshpetit commented on ESCPOS_NET/Emitters/BaseCommandEmitter/ImageCommands.cs:
```diff
@@ -118,5 +123,23 @@ public virtual byte[] PrintImage(byte[] image, bool isHiDPI, bool isLegacy = fal
+            var memoryStream = new MemoryStream();
+            image.SaveAsPng(memoryStream);
+            var bytes = memoryStream.ToArray();
+            memoryStream.Dispose();
```
In Java, there is garbage collection that can usually handle such cleanup. Just curious how this works in C#.

nickcharlton replied:
My understanding is that—as we're explicitly working with memory here—we need to make sure we dispose of the intermediary object …but I might be wrong!

nickcharlton committed:
We need to print it as an image

lukevp commented:
@nickcharlton  I merged another PR which bumped imagesharp, could you resolve the merge conflicts?

This is really cool, are you able to set widths and auto-wrap text to ensure it doesn't overflow the printer? or do you just have to rasterize each line individually?

nickcharlton committed:
Use ImageSharp.Drawing 1.0.0-beta14

nickcharlton commented:
I've just rebased to get up with those changes. I also added a commit to bump `ImageSharp.Drawing` to the latest beta as there were a few changes there too.

I think we probably could. Could you point me to where I could find out the widths available to us? I'm thinking the nicest solution here is to default to that but still allow overriding.

Also, do you have any thoughts on how we could go about testing this too?

lukevp commented:
I think everything you mentioned makes sense!  There's not really any automated testing in this project, just a console test, since most of the tests require validation on physical hardware to interpret the commands anyway.  I think that would be fine here as well, I don't think you need to have automated tests, just a print command in the console test that will print some rasterized stuff.

For default widths, I don't know what the defaults are, I think it would depend on the printer DPI and how wide it can print.  You could just have a conservative setting that works on most epson printers out of the box for a standard receipt size, and then another for the common smaller paper size.  I have printers of both sizes that I could test this on.

Regarding system.drawing.common, do you think we could use that instead of imagesharp? I know I picked imagesharp at the onset of the project instead of system.drawing.common for some compatibility reasons, but I'm not certain why.  Is it in net standard 2.0?

I've added immediate network printing support as it's an often-requested feature, and am releasing version 3.0, can you update this PR again? my apologies.  I would love to bring this in to version 3.1.

lukevp commented:
font wise, I think a public domain font would be best, if such a thing exists, since the MIT license may conflict with anything stricter than public domain (eg. I don't know how MIT and CC licenses play out together).

lukevp commented:
Do you use this for a line of text at a time, and wrap it manually somehow? or don't wrap?

djnuller commented on ESCPOS_NET/Emitters/BaseCommandEmitter/ImageCommands.cs:
```diff
@@ -119,5 +123,23 @@ public virtual byte[] PrintImage(byte[] image, bool isHiDPI, bool isLegacy = fal
+            var bytes = memoryStream.ToArray();
+            memoryStream.Dispose();
+
+            return PrintImage(bytes, true);
```
I would pass in the isHDPI.

raphac commented:
Hello
I would also be interested in the feature. What is the status here?

igorocampos commented:
> Hello I would also be interested in the feature. What is the status here?

Still with merge conflicts, and there's one open comment
